pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go

```diff
@@ -62,6 +61,20 @@ func init() {
 					}
 				}),
 			),
+			// The transition from MERGED to WRITE_ONLY must precede index validation.
+			// In MERGE_ONLY and MERGED, the index receives writes, but writes do not
+			// enforce uniqueness (they don't use CPut, see ForcePut). In WRITE_ONLY,
+			// the index receives normal writes. Only once writes are enforcing the
+			// uniqueness constraint for new can we validate that the index does
+			// indeed represent a valid uniqueness constraint over all rows.
+			to(scpb.Status_WRITE_ONLY,
+				emit(func(this *scpb.PrimaryIndex) *scop.MakeMergedIndexWriteOnly {
+					return &scop.MakeMergedIndexWriteOnly{
+						TableID: this.TableID,
+						IndexID: this.IndexID,
+					}
+				}),
+			),
```

Comment:
This is a big change! I previously don't know that an index that is in MERGE_ONLY or MERGED can still receive writes. What is the concern of checking uniqness of writes during merging previously (when we transition the primary index to WRITE_ONLY before merging with the temporary index)?

Reply:
While the index is in `MERGING` in the table descriptor, write to it use `Put` instead of `CPut`. See `ForcePut` on `catalog.Index`. 

Consider the following timeline:

create table t (i int primary key, j int);
insert into t values (1, 1);
create unique index idx on t(j);

t1: index merging concludes
at this point the unique index contains:
`1: 1`

t2: validation begins by reading a snapshot at t2 to count that the number of rows in the unique index is equal to the number of rows in the primary index
t3: `insert into t values (2, 1)` -- this will lead to the unique index now only containing `1: 2` instead of `1: 1` and the force from `MERGING` means the write will succeed.
t4: Validation completes successfully and the index is moved to public

At this point we have a corrupt unique secondary index.

If, however, we moved from `MERGING` to `WRITE_ONLY` before proceeding with validation, the insert operation would fail and the index will remain consistent.